tests/run-make/split-debuginfo/Makefile

```diff
@@ -11,23 +13,77 @@ off:
 	[ ! -d $(TMPDIR)/foo.dSYM ]
 
 # Packed by default, but only if debuginfo is requested
-packed:
-	rm -rf $(TMPDIR)/*.dSYM
-	$(RUSTC) foo.rs
-	[ ! -d $(TMPDIR)/foo.dSYM ]
-	rm -rf $(TMPDIR)/*.dSYM
-	$(RUSTC) foo.rs -g
+packed: packed-remapped-scope packed-remapped-wrong-scope
+
+# - Debuginfo in binary file
+# - `.o` deleted
+# - `.dSYM` present
+# - in binary, paths from `N_SO` (source files) and `N_OSO` (object files) shouldn be remapped
```

Comment:
Hey, @michaelwoerister. I've move <https://github.com/rust-lang/rust/pull/118518#pullrequestreview-1821356351> to this thread so we can follow the discussion better.

> I'm not quite clear yet on the situation on macOS. Do you have a link to some documentation on the various SO symbols?

It's also unclear to me. I have no idea if there is an official doc for Mach-O debug symbols. Here is what I've found:

* [Apple's “Lazy” DWARF Scheme](https://wiki.dwarfstd.org/Apple%27s_%22Lazy%22_DWARF_Scheme.md)
* [Apple's Linker & Deterministic Builds](https://milen.me/writings/apple-linker-ld64-deterministic-builds-oso-prefix/)
* [SymbolTable.pm](https://opensource.apple.com/source/cctools/cctools-973.0.1/tests/include/MachO/SymbolTable.pm) from old Apple open source header (search for `N_SO` and `N_OSO`)
* [ArchiveWriter.cpp](https://github.com/llvm/llvm-project/blob/e81c981fe35483a4b7474e110ff1b5d02bc2cb00/llvm/lib/Object/ArchiveWriter.cpp#L738-L751) from LLVM
* [My own experiments on it](https://github.com/rust-lang/rust/issues/116948#issuecomment-1793617018)

Reply:
I'll try to read up on those links by the end of the week. If we can't get to a clear conclusion, I think we should merge the PR nonetheless, for the improved behavior it provides on Linux.

Reply:
OK, from reading through those links it sounds to me like:

- SO symbols we can handle fine by remapping the paths generated by debuginfo.
- OSO symbols are added by the linker. They are not needed after dsymutil has built the dSYM bundle. So it would be fine to remove them after building the dSYM bundle.

I think the best option would be to use the `-oso_prefix` linker option. My guess is that will not be a problem even on older macOS versions. But I don't know if we have a minimum supported XCode version (in addition to a minimum support macOS version).

Other than that, rewriting/removing the OSO symbols in a postprocessing step seems to be our only option.

Reply:
@weihanglo, do you know how `-Cstrip=debuginfo` behaves on macOS? Does that remove the OSO symbols?

Reply:
It runs `strip -S` on macOS and removes OSO symbols successfully, and the `.dSYM` bundle can successfully find sources for a simple `hello-world.rs` example.

Should we always pass `-Cstrip=debuginfo` to deal with OSO symbols? I am not sure…

Reply:
It looks like `-Cstrip=debuginfo` is [now the default](https://github.com/rust-lang/cargo/pull/13257) unless debuginfo is requested. We could also do a similar thing for when trim-paths is enabled.

Reply:
That sounds good, but only when Cargo is involved. OSO is still there if using rustc directly.

(granted, people invoking rustc directly should be able to set `-Cstript=debuginfo`)

Reply:
or do you mean that rustc should implicitly strip debuginfo under some trim-paths circumstances?

Reply:
I was more thinking about the Cargo case. For direct rustc invocations, the defaults are less important, I think, because that's more of an advanced use case. The test cases here could simulate what we expect Cargo to do.